### PR TITLE
Throw if `document` is missing by the time invokeGuardedCallbackDev runs

### PR DESCRIPTION
**Pull request by gaearon**
The case described in https://github.com/facebookincubator/create-react-app/issues/3482 is pretty unintuitive.

This attempts to provide a better warning message when this happens. I opted not to modify the behavior since it's going to be pretty fragile and will probably break down the line anyway. Just printing a warning early.

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOM-test.js [hunk 1]
```diff
@@ -372,4 +372,45 @@ describe('ReactDOM', () => {
...
+    }
+    const container = document.createElement('div');
+    const instance = ReactDOM.render(<App />, container);
+    const documentDescriptor = Object.getOwnPropertyDescriptor(
+      global,
+      'document',
+    );
+    try {
+      // Emulate jsdom environment cleanup.
+      // This is rouhly what happens if the test finished and then
```
Comment (nhunzaker) [inline packages/react-dom/src/__tests__/ReactDOM-test.js]: Roughly

**Comment by sophiebits** [comment]
Is this the only `document` that can be missing? How sure are you that we don't (and won't) run into similar trouble in other paths?

**Comment by gaearon** [comment]
>Is this the only document that can be missing?

Maybe not. But if it's missing, it definitely *will* crash just a few lines below when we use `document.createEvent`. So warning never hurts.

It's likely (some? all?) other jsdom globals will be missing too. But I figured detecting this one is enough. Since we are the ones using it just below. I could also detect missing `window` but since the original report mentioned `document` I think it might be enough. We could later expand the check if we get similar reports mentioning `window` (that somehow still have `document` defined).

>How sure are you that we don't (and won't) run into similar trouble in other paths?

Not quite sure what you mean. If `document` is missing we'll fail in this very method before we do anything interesting. Since we use `document.createEvent`.

**Comment by sophiebits** [comment]
I meant is this always the outermost wrapper? We never use document anywhere else where it's not wrapped in this?

**Comment by gaearon** [comment]
It does seem so. It's used in scheduler for committing host effects, lifecycles, and the main work loop. In other words, everywhere that boundaries need to catch.

**Comment by gaearon** [comment]
I have changed it to throw. It turns out that an asynchronous error like this actually causes Jest to fatal, and the warning doesn't get a chance to get printed. So now I'm just throwing preemptively with a better message.

**Pull request merged**